psydac/linalg/tests/test_tosparse_parallel.py

```diff
@@ -0,0 +1,226 @@
+@pytest.mark.parametrize( 'P2', [True] )
+@pytest.mark.parallel
+
+def test_block_linear_operator_parallel_dot( dtype, n1, n2, p1, p2, P1, P2 ):
+    # Define a factor for the data
+    if dtype == complex:
+        factor = 1j
+    else:
+        factor = 1
+
+    # set seed for reproducibility
+    np.random.seed(n1*n2*p1*p2)
+
+    from mpi4py import MPI
+
+    comm = MPI.COMM_WORLD
+    rank = comm.Get_rank()
+    size = comm.Get_size()
+
+    D = DomainDecomposition([n1,n2], periods=[P1,P2], comm=comm)
+
+    # Partition the points
+    npts = [n1,n2]
+    global_starts, global_ends = compute_global_starts_ends(D, npts)
+
+    cart = CartDecomposition(D, npts, global_starts, global_ends, pads=[p1,p2], shifts=[1,1])
+
+    # Create vector space, stencil matrix, and stencil vector
+    V = StencilVectorSpace( cart, dtype=dtype )
+
+    v0arr, v0 = create_equal_random_arrays(V, seedv=4568)
+    v0arr = v0arr[0].flatten()
+
+    # Create and Fill Block objects
+    W = BlockVectorSpace(V, V)
+
+    v1arr1, v1 = create_equal_random_arrays(W, seedv=4568)
+    v1arr = []
+    for i in v1arr1:
+        aux = i.flatten()
+        for j in aux:
+            v1arr.append(j)
+ 
+    # Create random matrix 
+    N1 = StencilMatrix( V, V )
+    N2 = StencilMatrix( V, V )
+    N3 = StencilMatrix( V, V )
+    N4 = StencilMatrix( V, V )
+
+    for k1 in range(-p1,p1+1):
+        for k2 in range(-p2,p2+1):
+            N1[:,:,k1,k2] = factor*np.random.random()
+            N2[:,:,k1,k2] = factor*np.random.random()
+            N3[:,:,k1,k2] = factor*np.random.random()
+            N4[:,:,k1,k2] = factor*np.random.random()
+
+    K = BlockLinearOperator( W, W )
+
+    K[0,0] = N1
+    K[0,1] = N2
+    K[1,0] = N3
+    K[1,1] = N4
+
+    #####
+    #Testing tosparse() for LinearOperators with domain being StencilVectorSpace
+    
+    #We take a power of the Stencil Matrix N1
+    N1P = N1**3
+    #We turn our PowerLinearOperator into a sparse matrix, and then into a numpy array
+    N1arrloc = N1P.tosparse().toarray()
+    #We get the global matrix of the PowerLinearOperator
+    N1arr = np.zeros(np.shape(N1arrloc), dtype=dtype) 
+    comm.Allreduce(N1arrloc, N1arr, op=MPI.SUM)
+    
+    assert isinstance(N1P, PowerLinearOperator)
+    assert isinstance(N1P.domain, StencilVectorSpace)
+    compare_arrays(N1P.dot(v0), N1arr.dot(v0arr), rank)
+    
+    #####
+    #Testing tosparse() for LinearOperators with domain being BlockVectorSpace
+    
+    #We take a power of the BlockLinearOperator
+    KP = K**3
+    #We turn our PowerLinearOperator into a sparse matrix, and then into a numpy array
+    KLarrloc = KP.tosparse().toarray()
+    #We get the global matrix of the operator
+    KLarr = np.zeros(np.shape(KLarrloc), dtype=dtype)
+    comm.Allreduce(KLarrloc, KLarr, op=MPI.SUM)
+
+    assert isinstance(KP, PowerLinearOperator)
+    assert isinstance(KP.domain, BlockVectorSpace)
+    compare_arrays(KP.dot(v1), KLarr.dot(v1arr), rank)
```

Comment:
Since we assemble the dense matrix, couldn't we compare the matrix arrays directly?